python/lsst/pipe/base/connections.py

```diff
@@ -0,0 +1,389 @@
+        # python documentation on metaclasses
+        super().__init__(name, bases, dct)
+
+
+class QuantizedConnection(SimpleNamespace):
+    """A Namespace to map defined variable names of connections to their `lsst.daf.buter.DatasetRef`s
+
+    This class maps the nameds used to define a connection on a PipelineTaskConnectionsClass to
+    the corresponding `lsst.daf.butler.DatasetRef`s provided by a `lsst.daf.butler.Quantum` instance.
+    This will be a quantum of execution based on the graph created by examining all the connections
+    defined on the `PipelineTaskConnectionsClass`.
+    """
+    def __init__(self, **kwargs):
+        # Create a variable to track what attributes are added. This is used later when
+        # iterating over this QuantizedConnection instance
+        object.__setattr__(self, "_attributes", set())
+
+    def __setattr__(self, name: str, value: typing.Union[DatasetRef, typing.List[DatasetRef]]):
+        # Capture the attribute name as it is added to this object
+        self._attributes.add(name)
+        super().__setattr__(name, value)
+
+    def __iter__(self) -> typing.Generator[typing.Tuple[str, typing.Union[DatasetRef,
+                                           typing.List[DatasetRef]]], None, None]:
+        """Make an Iterator for this QuantizedConnection
+
+        Iterating over a QuantizedConnection will yield a tuple with the name of an attribute and
+        the value associated with that name. This is similar to dict.items() but is on the
+        namespace attributes rather than dict keys.
+        """
+        yield from ((name, getattr(self, name)) for name in self._attributes)
+
+    def keys(self) -> typing.Generator[str, None, None]:
+        """Returns an iterator over all the attributes added to a QuantizedConnection class
+        """
+        yield from self._attributes
+
+
+class InputQuantizedConnection(QuantizedConnection):
+    pass
+
+
+class OutputQuantizedConnection(QuantizedConnection):
+    pass
+
+
+class DeferredDatasetRef(namedtuple("DeferredDatasetRef", "datasetRef")):
+    """Class which denotes that a datasetRef should be treated as deferred when interacting with the butler
+
+    Parameters
+    ----------
+    datasetRef : `lsst.daf.butler.DatasetRef`
+        The `lsst.daf.butler.DatasetRef` that will be eventually used to resolve a dataset
+    """
+    __slots__ = ()
+
+
+class PipelineTaskConnections(metaclass=PipelineTaskConnectionsMetaclass):
+    """PipelineTaskConnections is a class used to declare desired IO when a PipelineTask is run by an
+    activator
+
+    PipelineTaskConnection classes are created by declaring class attributes of types defined in
+    lsst.pipe.base.connectionTypes and are listed as follows:
+
+    * InitInput - Defines connections in a quantum graph which are used as inputs to the __init__ function
+                  of the PipelineTask corresponding to this class
```

Comment:
The lines should not be indented or they will become indented in sphinx output too. I think you want this:

    * InitInput - Defines connections in a quantum graph which are used as inputs to the __init__ function
      of the PipelineTask corresponding to this class

(but also properly wrapped at 79 characters).

You can also try try definition list syntax: http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#definition-lists